### DISCUSSION TITLE
Any latest performance report available?

### DISCUSSION THREAD

**Discussion by hanijohn** [fvutils/pyvsc#192] [Q&A]
I work on designs with a large set of constraints using arrays, so my query is based on below comment from the paper (3yrs old)
https://github.com/fvutils/pyvsc/blob/master/doc/papers/PyVSC_Randomization_Constraints_Coverage_in_Python_final.pdf

"A natural next step is to re-implement the core constraint and
coverage data mode in C++ and expose it as a Python extension.
Early experiments with this approach have shown significant
performance improvements are possible in cases where the
number of variables is significant."

How is the performance of latest pyvsc? Any comparative study available with the constraint solver in Simulator?

**Comment by mballance** [comment]
Hi @hanijohn,
    As you've observed, projects to make significant solver performance improvements are taking longer than I had originally expected. As of now, none of those improvement projects are available for use. However, work on the next iteration of PyVSC is very much underway.

- The next major release of PyVSC will separate the front end (the language-like features for capturing constraints) from the back-end, and supports a more concise way to capture data structures using Python 'dataclasses' ([PyVSC Dataclasses](https://github.com/vsc-tools/pyvsc-dataclasses))
- Making performance comparisons is challenging with the current PyVSC since it's Python-only. The [VSC Systest](https://github.com/vsc-tools/vsc-systest) project converts testcases captured in Python to SystemVerilog and other formats to support performance comparisons.
- Back-end support for constraints solving will be provided by the [VSC-DM](https://github.com/vsc-tools/vsc-dm) project (data model) and [VSC-Solvers](https://github.com/vsc-tools/vsc-solvers) projects
  - Enable the use of multiple solvers depending on the type of problem
  - Provides solver infrastructure that can be used in C++ and other language, in addition to Python

I'm not in a position to provide an estimate for when some of these new enhancements will be available for use, but the work certainly well underway.

Best Regards,
Matthew
